latch/registry/utils.py

```diff
@@ -0,0 +1,343 @@
+        )
+
+    if "array" in registry_type:
+        sub_type = registry_type["array"]
+
+        if not type(python_literal) is list:
+            raise ValueError(
+                "unable to convert non-list python literal to registry array literal"
+            )
+
+        return [to_registry_literal(literal, sub_type) for literal in python_literal]
+
+    if not "primitive" in registry_type:
+        raise ValueError(f"malformed registry type: {registry_type}")
+
+    primitive = registry_type["primitive"]
+
+    value = None
+    if primitive in ["string", "date", "datetime"]:
+        if not isinstance(python_literal, str):
+            raise ValueError(
+                f"cannot convert non-string python literal to {primitive} registry"
+                " literal"
+            )
+        value = python_literal
+    elif primitive == "integer":
+        if not isinstance(python_literal, int):
+            raise ValueError(
+                "unable to convert non-int python literal to registry int literal"
+            )
+        value = int(python_literal)
+    elif primitive == "number":
+        if not (isinstance(python_literal, int) or isinstance(python_literal, float)):
+            raise ValueError(
+                "unable to convert non-numeric python literal to registry number"
+                " literal"
+            )
+        value = float(python_literal)
+    elif primitive == "boolean":
+        if not isinstance(python_literal, bool):
+            raise ValueError(
+                "unable to convert non-boolean python literal to registry boolean"
+                " literal"
+            )
+        value = python_literal
+    elif primitive == "null":
+        if not python_literal is None:
+            raise ValueError(
+                "unable to convert non-None python literal to registry null literal"
+            )
+    elif primitive == "enum":
+        members = registry_type["members"]
+        if isinstance(python_literal, Enum):
+            python_literal = python_literal.name
+        elif not isinstance(python_literal, str):
+            python_literal = str(python_literal)
+        if not python_literal in members:
+            raise ValueError(
+                f"unable to convert {python_literal} to registry enum with members"
+                f" {', '.join(members)}"
+            )
+        value = python_literal
+    elif primitive == "link":
+        from latch.registry.record import Record
+
+        if not isinstance(python_literal, Record):
+            raise ValueError(
+                "cannot convert non-record python literal to registry link"
+            )
+        value = {"sampleId": python_literal.id}
+    elif primitive == "blob":
+        if not (
+            isinstance(python_literal, LatchFile)
+            or isinstance(python_literal, LatchDir)
+        ):
+            raise ValueError("cannot convert non-blob python literal to registry blob")
+
+        node_id = execute(
```

Comment:
we probably don't want to marshal to-from path here and just deal with IDs instead

instead we should:
1. convert to `LatchFile` only when needed
2. resolve paths to IDs as soon as they are written (as part of new value validation)

same as console does